### PR TITLE
Recompute steps_per_epoch after each epoch in traingin_generator

### PR DESCRIPTION
**Pull request by vikua**
### Summary
keras.util.Sequence gives great flexibility to implement user's own
train sequences, but number of steps per epoch is computed only once at
the beginning of the training. This makes is hard to change sequence
lenghts (like as a result of changed batch size) between different
epochs. Forcing training_generator to recompute steps_per_epoch after
each epoch to avoid that.

### Related Issues
https://github.com/keras-team/keras/issues/10143
https://github.com/keras-team/keras/issues/13038

### PR Overview
Recompute `steps_per_epoch` after each epoch in `training_generator.py::fit_generator`
I didn't add any tests for now cause I'm not aware of the gotchas and looking for a feedback first. 
 

- [x] This PR requires new unit tests [y/n] (make sure tests are included)
- [ ] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [ ] This PR is backwards compatible [y/n]
- [ ] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### CODE REVIEW

**Comment by vikua** [comment]
More investigation on this in the issue #13038 

In general looks like `Sequence.on_epoch_end` call is sort of broken right now.
For main thread generator - it just never called.
For `OrderedEnqueuer` it called not right after the epoch has finished but in more unpredictable way.
For example I ran the test with prints in `on_epoch_end` and before calling `next(output_generator)` in training loop. Initial batch size is 3, batch size gets increased by 2 every time. Sequence length is 20. Here is the test (also present in the PR):  
```
    tracker_cb = TrackerCallback()
    val_seq = RandomSequence(4)
    train_seq = IncreaseBatchSizeRandomSequence(3, 20)
    out = model.fit_generator(generator=train_seq,
                              epochs=5,
                              initial_epoch=0,
                              validation_data=val_seq,
                              validation_steps=3,
                              max_queue_size=1,
                              callbacks=[tracker_cb])
```

The output in this case: 
```
Epoch 1/5
Grapping next item, epoch 0, batch 0
Grapping next item, epoch 0, batch 1
Grapping next item, epoch 0, batch 2
Grapping next item, epoch 0, batch 3
Grapping next item, epoch 0, batch 4
Grapping next item, epoch 0, batch 5
Grapping next item, epoch 0, batch 6
Epoch 2/5
Grapping next item, epoch 1, batch 0
Calling on_epoch_end
Grapping next item, epoch 1, batch 1
Grapping next item, epoch 1, batch 2
Grapping next item, epoch 1, batch 3
Grapping next item, epoch 1, batch 4
Calling on_epoch_end
Grapping next item, epoch 1, batch 5
Grapping next item, epoch 1, batch 6
Epoch 3/5
Grapping next item, epoch 2, batch 0
Calling on_epoch_end
Grapping next item, epoch 2, batch 1
Grapping next item, epoch 2, batch 2
Epoch 4/5
Grapping next item, epoch 3, batch 0
Calling on_epoch_end
Grapping next item, epoch 3, batch 1
Grapping next item, epoch 3, batch 2
Calling on_epoch_end
Epoch 5/5
Grapping next item, epoch 4, batch 0
Grapping next item, epoch 4, batch 1
Calling on_epoch_end
```

`on_epoch_end` is called not after 1st epoch but after first item of the 2nd epoch was grabbed. 2nd epoch is longer then expected (7 batches vs 4 expected). Also a lot of `on_epoch_end` happen not when they are expected. 

This PR fixes that issue.

File: keras/utils/data_utils.py [hunk 1]
```diff
@@ -561,18 +561,23 @@ def _wait_queue(self):
 
     def _run(self):
         """Submits request to the executor and queue the `Future` objects."""
-        sequence = list(range(len(self.sequence)))
-        self._send_sequence()  # Share the initial sequence
         while True:
+            # get new sequence indexes for each epoch
+            sequence = list(range(len(self.sequence)))
+            # Update pool with new sequence
+            self._send_sequence()
```
Comment (vikua) [inline keras/utils/data_utils.py]: We need to update sequence and sent to the pool at the beginning of each epoch. This is done to make sure correct indexes are used in case if Sequence changes it's length.

File: keras/utils/data_utils.py [hunk 2]
```diff
@@ -561,18 +561,23 @@ def _wait_queue(self):
...
-                for i in sequence:
+                for i, item_idx in enumerate(sequence):
                     if self.stop_signal.is_set():
                         return
-                    future = executor.apply_async(get_index, (self.uid, i))
-                    future.idx = i
+                    future = executor.apply_async(get_index, (self.uid, item_idx))
+                    future.idx = item_idx
+                    # last is True if item is last in current epoch
+                    future.last = i == len(sequence) - 1
```
Comment (vikua) [inline keras/utils/data_utils.py]: I tried to call `self.sequence.on_epoch_end` before `self._wait_queue()` call. This solves the issue of correct sequence length, but there are issues with the fact that sequence is shared between current thread and executor - now we can change state of the sequence before last future  from previous epochs has been done. This means we call `on_epoch_end` to start a new epoch, but executor still processes some of the `get_index` functions from previous epochs. 

That's why I added a flag to the future, which marks it as last in epoch. It will be true for last future in sequence for each epoch.  
This flag is used in `get` method to trigger `on_epoch_end` once executor not only excepted a task, but actually processed it.

**Comment by vikua** [comment]
@zachmayer fyi

File: keras/utils/data_utils.py (hunk 1 above)
Comment (zachmayer) [reply keras/utils/data_utils.py]: Maybe add a comment to the code saying this too?

**Comment by vikua** [comment]
Hi @fchollet, @Dref360, @gabrieldemarmiesse  very sorry for pinging you guys directly on this. This pr has been open for 10 days without a feedback, so I just wanted to get your attention to it.  
In my opinion this is really nice to have so let me know what you think!

File: keras/engine/training_generator.py [hunk 3]
```diff
@@ -252,11 +257,28 @@ def fit_generator(model,
                 if callbacks.model.stop_training:
                     break
 
+            if use_sequence_api and workers <= 0:
```
Comment (Dref360) [inline keras/engine/training_generator.py]: Should be == instead of <=

File: keras/utils/data_utils.py (hunk 2 above)
Comment (Dref360) [reply keras/utils/data_utils.py]: This adds a lot of complexity. If we wait for the queue to be empty, then there is no batch getting computed so it is safe to modify.

File: tests/keras/engine/test_training.py [hunk 4]
```diff
@@ -603,6 +630,134 @@ def gen_data(i):
...
+    out = model.fit_generator(generator=train_seq,
+                              epochs=5,
+                              initial_epoch=0,
+                              validation_data=val_seq,
+                              validation_steps=3,
+                              max_queue_size=1,
+                              callbacks=[tracker_cb])
+    assert tracker_cb.trained_epochs == [0, 1, 2, 3, 4]
+    assert tracker_cb.trained_batches == [
+        0, 1, 2, 3, 4, 5, 6,  # 1st epoch -> cail(20 / 3) = 7 batches
```
Comment (Dref360) [inline tests/keras/engine/test_training.py]: ceil*

**Comment by vikua** [comment]
@Dref360 yes, progress bar will be updated with new steps each epoch

File: keras/utils/data_utils.py (hunk 2 above)
Comment (vikua) [reply keras/utils/data_utils.py]: That's why I modify sequence after last batch is consumed as well - to make sure no tasks are inside of executer when we call `self.sequence.on_epoch_end`. 

https://github.com/vikua/keras/blob/vikua/seq-gen-steps_per_epoch/keras/utils/data_utils.py#L606-L610
As you can use  `self.sequence.on_epoch_end` is called only when consumed future is last in epoch, result is computed (after `future.get(timeout=30)`) and before notifying `self.queue.task_done` to make sure new epoch is not started yet inside of `_run`.

**Comment by vikua** [comment]
@Dref360 Regarding a prog bar, here is a code snippet: 

```
import numpy as np
import keras

from keras import Model
from keras.layers import Dense, Input
from keras.utils import Sequence
from keras.datasets import boston_housing


class BatchSizeSchedulerSequence(Sequence):
    def __init__(self, X, y, initial_batch_size=32):
        self.X = X
        self.y = y
        self.batch_size = initial_batch_size

    def __len__(self):
        new_len = int(np.ceil(len(self.X) / float(self.batch_size)))
        # print('BSS Len:', new_len, 'batch_size', self.batch_size)
        return new_len

    def __getitem__(self, idx):
        batch_x = self.X[idx * self.batch_size:(idx + 1) * self.batch_size]
        batch_y = self.y[idx * self.batch_size:(idx + 1) * self.batch_size]
        return np.array(batch_x), np.array(batch_y)

    def on_epoch_end(self):
        #print('Calling on_epoch_end')
        self.batch_size *= 2

(X_train, y_train), (X_test, y_test) = boston_housing.load_data()

inputs = Input(shape=(X_train.shape[1],))
outputs = Dense(64, activation='relu')(inputs)
outputs = Dense(1)(outputs)
model = Model(inputs, outputs)

model.compile(optimizer='adam', loss='mae')

train_seq = BatchSizeSchedulerSequence(X_train, y_train, initial_batch_size=2)
model.fit_generator(train_seq, workers=1, epochs=10, validation_data=(X_test, y_test))
```

and here is it's output: 
```
Epoch 1/10
202/202 [==============================] - 0s 2ms/step - loss: 20.8676 - val_loss: 7.8014
Epoch 2/10
101/101 [==============================] - 0s 2ms/step - loss: 5.9063 - val_loss: 6.3413
Epoch 3/10
51/51 [==============================] - 0s 3ms/step - loss: 5.3959 - val_loss: 5.3670
Epoch 4/10
26/26 [==============================] - 0s 4ms/step - loss: 5.0446 - val_loss: 5.3726
Epoch 5/10
13/13 [==============================] - 0s 8ms/step - loss: 4.7872 - val_loss: 5.2622
Epoch 6/10
7/7 [==============================] - 0s 14ms/step - loss: 4.6092 - val_loss: 5.3788
Epoch 7/10
4/4 [==============================] - 0s 25ms/step - loss: 4.4638 - val_loss: 5.2272
Epoch 8/10
2/2 [==============================] - 0s 52ms/step - loss: 4.7832 - val_loss: 5.2152
Epoch 9/10
1/1 [==============================] - 0s 101ms/step - loss: 4.7433 - val_loss: 5.2156
Epoch 10/10
1/1 [==============================] - 0s 101ms/step - loss: 4.7094 - val_loss: 5.2235

```

As you can see progbar is updated

File: keras/utils/data_utils.py (hunk 2 above)
Comment (Dref360) [reply keras/utils/data_utils.py]: Sorry I don't think I get it.

When we call `_wait_queue()`, it will block till the last `task_done` is called. At this point, there is no batch getting processed, the queue is empty and no new future is getting queued. 


At this point, it should be safe to call `on_epoch_end`. If it's not the case, maybe there is a bug in `wait_queue`?

Comment (vikua) [reply keras/utils/data_utils.py]: I'm not sure 100% why it happens (trying to figure out), but if I move `self.sequence.on_epoch_end` after `_wait_queue()` - tests I added here will fail. For example in `test_fit_generator_dynamic_size_sequence_with_workers` second epoch will always have same length as first epoch, whereas it should be shorter.

Comment (vikua) [reply keras/utils/data_utils.py]: I think it is happening because `on_epoch_end` is called in the separate thread, whereas main thread is still spinning. 
So basically main thread consumes last items from the queue, updates queue with `task_done` an immediately recomputes steps_per_epoch and goes to next `queue.get` call. At this point `OrderedEnqueuer._run` thread might not called `self.sequence.on_epoch_end` yet. 
That's why new epoch gets wrong `steps_per_epoch`.

So in order to guarantee `steps_per_epoch` recomputed _after_ `self.sequence.on_epoch_end ` call I moved it to the main thread.

Comment (vikua) [reply keras/utils/data_utils.py]: @Dref360 I did same but using simpler way - just calling `sequence.on_epoch_end` in training loop 
commit is 8b7da25930fb5ac6826d834d7d01b26cbdd56471

This does same as previous version, but I think some tests in `data_utils_test.py` related to `OrderedEnqueuer` will fail, cause it doesn't call `self.sequence.on_epoch_end ` now

Comment (vikua) [reply keras/utils/data_utils.py]: What do you think - should be keep `self.sequence.on_epoch_end` call in `OrderedEnqueuer` (but have slightly more complicated version), or have `generator.on_epoch_end` call in training loop and change some tests in `data_utils_test.py `?

Comment (vikua) [reply keras/utils/data_utils.py]: @Dref360 I found another way to pull correct sequence length from `OrderedEnqueuer._run` thread into the main thread. I added another commit d8c1ab4ab86150a787256f693986c5a9378c43ae with the change.
Basically I use separate queue with just 1 slot to push information from run thread to main thread. This queue acts as an exchange between those thread. Whenever run thread is done thing an epoch it puts new length into this queue. Main thread on the other side pulls that queue for new sequence length. 
This helps to keep `OrderedEnqueuer` making `self.sequence.on_epoch_end` call. 
Also this helps to ensure correct order of `update sequence -> get length` events. 

So basically there are three ways right now 
- use `last` flag on the future (like I did initially) and call on_epoch_end in main thread in `OrderedEnqueuer.get`
- call on_epoch_end in main thread in training loop (simple, but some tests will need changing)
- use single slot queue to exchange information between main thread and `OrderedEnqueuer._run` thread.

**Comment by Dref360** [comment]
I think using your third idea would be better and could be generalized to an observer pattern. In this branch, I used an Event, but it could also be generalized : https://github.com/keras-team/keras/compare/master...Dref360:my_tale?expand=1



IDEA: If we set steps_per_epoch to be a dynamic int that changes whenever we call on_epoch_end, it would be the best imo.

I'm in a hurry so if it's not clear do not bother. But I think that something like:

```python
observable_int = ObservableInt(steps_per_epoch)
enqueuer.register_on_epoch_end_callback(lambda enqueuer: observable_int.set_value(len(enqueuer.sequence)))

...
callbacks.set_params({
                    'epochs': epochs,
                    'steps': observable_int, # Where the value is a property that update itself.
                    'verbose': verbose,
                    'do_validation': do_validation,
                    'metrics': callback_metrics,
                })


```

Now whenever we call sequence.on_epoch_end(), the enqueuer would modify this value, this value will affect the ProgBar.

**Comment by vikua** [comment]
Got it. Do you want me to incorporate this Observer pattern in this branch? 
Idea seems to be very clear to me so I can do it today

**Comment by Dref360** [comment]
I'm going on vacation for a week so I won't be able to review. In the end, I think that we don't want to add too much complexity whenever possible hence why the observable pattern that inherits __int__ would be the best in my opinion.

**Comment by vikua** [comment]
@Dref360 I modified the code to use `threading.Event`. Thanks for the idea!

**Comment by vikua** [comment]
@Dref360  Regarding `ObserbableInt` - I have some doubts. The idea is basically to subclass Python `int` class and make it mutable, so that callbacks can get new value of `steps_per_epoch` automatically. 
Not sure subclassing builtin Python value types directly is a good idea.

This object will go to custom callbacks and might be confusing for their authors.

Although I can try of course.

**Comment by vikua** [comment]
Still thinking about this ObservableInt. You've said you guys don't want to add too much complexity. 
I'm looking into ways how to implement it with observable pattern and it looks like it introduces way more complexity. 
ObservableInt will help to avoid this call 

https://github.com/keras-team/keras/blob/babaec7ae817fc7d0ad3e3f3eeedbe5425f35534/keras/engine/training_generator.py#L276

because callbacks will get new value of steps_per_epoch automatically. 
That's it. Threading event and synchronization will still be needed to make sure next epoch doesn't start before ObservableInt is updated. 

On the other hand with ObservableInt approach it's needed to implement mutable int itself. Also the fact that callbacks get new value implicitly - also introduces some complexity. 

not sure what is more complex - extending python value type and adding observable pattern or just update callbacks with new value of `steps_per_epoch`

**Comment by Dref360** [comment]
LGTM! I'll request the approval of @fchollet because it's an important change.

File: keras/engine/training_generator.py [hunk 5]
```diff
@@ -257,6 +262,25 @@ def fit_generator(model,
             if callbacks.model.stop_training:
                 break
 
+            if use_sequence_api and workers == 0:
```
[the remark this answers is not in the document]
Comment (vikua) [reply keras/engine/training_generator.py]: Previously - no. With this change - yes

[the remark this answers is not in the document]
Comment (vikua) [reply keras/engine/training_generator.py]: We do, in OrderedEnqueuer, like before. 

We never called on_epoch_end when workers == 0 before. This change adds that

[the remark this answers is not in the document]
Comment (mds47) [reply keras/engine/training_generator.py]: @fchollet does this make sense? Anything else we should check in order to merge?

**Comment by zachmayer** [comment]
@fchollet @Dref360 Do you think this PR is ready to merge?
